packages/cli/src/tools/metroPlatformResolver.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * This is an implementation of a metro resolveRequest option which will remap react-native imports
+ * to different npm packages based on the platform requested.  This allows a single metro instance/config
+ * to produce bundles for multiple out of tree platforms at a time.
+ *
+ * @param platformImplementations
+ * A map of platform to npm package that implements that platform
+ *
+ * Ex:
+ * {
+ *    windows: 'react-native-windows'
+ *    macos: 'react-native-macos'
+ * }
+ */
+// @ts-ignore - no typed definition for the package
+import {resolve} from 'metro-resolver';
+
+export function reactNativePlatformResolver(platformImplementations: {
+  [platform: string]: string;
+}) {
+  return (
+    context: any,
+    _realModuleName: string,
+    platform: string,
+    moduleName: string,
+  ) => {
+    let backupResolveRequest = context.resolveRequest;
+    delete context.resolveRequest;
+
+    try {
+      let modifiedModuleName = moduleName;
+      if (platformImplementations[platform]) {
+        if (moduleName === 'react-native') {
+          modifiedModuleName = platformImplementations[platform];
+        } else if (moduleName.startsWith('react-native/')) {
+          modifiedModuleName = `${
+            platformImplementations[platform]
+          }/${modifiedModuleName.slice('react-native/'.length)}`;
+        }
+      }
+      let result = resolve(context, modifiedModuleName, platform);
+      context.resolveRequest = backupResolveRequest;
+      return result;
+    } catch (e) {
+      context.resolveRequest = backupResolveRequest;
```

Comment:
We could use `finally` to bring back original resolveRequest once. Btw, is this really the official way to alias modules?

Reply:
Normally we alias modules using extraNodeModules.  We have to use the resolveRequest option here, because we want to alias the module differently based on which platform is being bundled, so you can run a single metro config and serve up any of the platforms.